topeft/modules/corrections.py

```diff
@@ -718,143 +966,184 @@ def AttachMuonSF(muons, year):
+        new_err = SFevaluator['MuonSF_{year}_er'.format(year=year)](eta,pt)
+        new_up = new_sf + new_err
+        new_do = new_sf - new_err
+
+    elif dt_era == "Run3":
+        loose_sf_flat = ak.where(
+            ~pt_mask,
+            1,
+            ceval["NUM_LooseID_DEN_TrackerMuons"].evaluate(abseta_flat, pt_flat_loose, "nominal")
+        )
+        loose_err_flat = ak.where(
+            ~pt_mask,
+            1,
+            np.sqrt(
+                ceval["NUM_LooseID_DEN_TrackerMuons"].evaluate(abseta_flat, pt_flat_loose, "syst") * ceval["NUM_LooseID_DEN_TrackerMuons"].evaluate(abseta_flat, pt_flat_loose, "syst") + ceval["NUM_LooseID_DEN_TrackerMuons"].evaluate(abseta_flat, pt_flat_loose, "stat") * ceval["NUM_LooseID_DEN_TrackerMuons"].evaluate(abseta_flat, pt_flat_loose, "stat")
+            )
+        )
+        loose_sf  = ak.unflatten(loose_sf_flat, ak.num(pt))
+        loose_err = ak.unflatten(loose_err_flat, ak.num(pt))
+        loose_up = loose_sf + loose_err
+        loose_do = loose_sf - loose_err
+
+        ##Not available yet for Run3
+        #reco_loose_sf_flat = ak.where(
+        #    ~pt_mask,
+        #    1,
+        #    ceval["NUM_LooseID_DEN_genTracks"].evaluate(abseta_flat, pt_flat_loose, "nominal")
+        #)
+        #reco_loose_err_flat = ak.where(
+        #    ~pt_mask,
+        #    1,
+        #    ceval["NUM_LooseID_DEN_genTracks"].evaluate(abseta_flat, pt_flat_loose, "syst") * ceval["NUM_LooseID_DEN_genTracks"].evaluate(abseta_flat, pt_flat_loose, "syst") + ceval["NUM_LooseID_DEN_genTracks"].evaluate(abseta_flat, pt_flat_loose, "stat") * ceval["NUM_LooseID_DEN_genTracks"].evaluate(abseta_flat, pt_flat_loose, "stat")
+        #)
+
+        #reco_sf = ak.unflatten(reco_sf_flat, ak.num(pt))
+        #reco_err = ak.unflatten(reco_err_flat, ak.num(pt))
+        #reco_up = reco_sf + reco_err
+        #reco_do = reco_sf - reco_err
+        #reco_loose_sf = ak.unflatten(reco_loose_sf_flat, ak.num(pt))
+        #reco_loose_err = ak.unflatten(reco_loose_err_flat, ak.num(pt))
+        #reco_loose_up = reco_loose_sf + reco_loose_err
+        #reco_loose_do = reco_loose_sf - reco_loose_err
+
+    #Run2+Run3 implementation
+    muons['sf_nom_2l_muon'] = new_sf * reco_loose_sf * iso_sf
+    muons['sf_hi_2l_muon']  = new_up * reco_loose_up * iso_up
+    muons['sf_lo_2l_muon']  = new_do * reco_loose_do * iso_do
+    muons['sf_nom_3l_muon'] = new_sf * reco_loose_sf
+    muons['sf_hi_3l_muon']  = new_up * reco_loose_up * iso_up
+    muons['sf_lo_3l_muon']  = new_do * reco_loose_do * iso_do
     muons['sf_nom_2l_elec'] = ak.ones_like(new_sf)
     muons['sf_hi_2l_elec']  = ak.ones_like(new_sf)
     muons['sf_lo_2l_elec']  = ak.ones_like(new_sf)
     muons['sf_nom_3l_elec'] = ak.ones_like(new_sf)
     muons['sf_hi_3l_elec']  = ak.ones_like(new_sf)
     muons['sf_lo_3l_elec']  = ak.ones_like(new_sf)
 
-def AttachElectronSF(electrons, year):
+def AttachElectronSF(electrons, year, looseWP="wp90noiso"):
```

Comment:
Is the default `looseWP` used?

Reply:
for Run3 yes, it is used in the analysis processor. Btw, maybe we should clearly define it in the `analysis_processor` and propagate it to this function. What do you think?

Reply:
Yes, let's try to avoid having a default value, and throw and exception if one is not provided.

Reply:
done